### PR TITLE
BUG: Fix inconsistency between the shape properties of SparseSeries and SparseArray (#21126)

### LINKED ISSUE #21126: SparseArray compatibility issues in e.g. sklearn. Series.shape and SparseArray.shape output not consistent.
#### Code Sample

```python
import pandas as pd
import numpy as np
spdf = pd.DataFrame(np.random.rand(5,5)>0.7).astype(float).to_sparse(fill_value=0)
spdf[1].shape[0] == spdf[1].values.shape[0]
```

#### Problem description
Using packages that internally translate pandas dataframes to arrays (like sklearn) requires vectors and matrices to be specific sizes. Using sparse data matrices in pandas produces unpredictable results as internals of sklearn will translate pandas to SparseArray arrays and check sizes. The size of `spdf[1].values.shape` is not guaranteed to have the expected shape output.

Also depending on the order of columns selection (after or before applying values) gives different results as the output of `spdf.values` is a `numpy.ndarray`.

#### Expected Output
I would expect the last statement in the code above to return True for all vectors in spdf.
I'm not sure this is a bug or intended feature but it struck me as weird as well as inconvenient when working with different libraries.
I encountered this running sklearn.linear_model.Lasso with a sparse target vector.

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Linux
OS-release: 4.6.0-040600-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.23.0
pytest: None
pip: 10.0.1
setuptools: 39.1.0
Cython: None
numpy: 1.14.2
scipy: 1.0.1
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: None
patsy: 0.5.0
dateutil: 2.7.2
pytz: 2018.4
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.2.2
openpyxl: 2.4.9
xlrd: 1.1.0
xlwt: None
xlsxwriter: None
lxml: 4.1.1
bs4: 4.6.0
html5lib: 0.9999999
sqlalchemy: 1.1.15
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by nprad**
- [x] closes #21126 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by nprad** [comment]
@TomAugspurger I made the changes you suggested. Let me know if there are any changes you need me to make.

File: pandas/core/sparse/array.py [hunk 1]
```diff
@@ -290,7 +290,9 @@ def __reduce__(self):
         """Necessary for making this object picklable"""
         object_state = list(np.ndarray.__reduce__(self))
         subclass_state = self.fill_value, self.sp_index
-        object_state[2] = (object_state[2], subclass_state)
+        object_state[2] = list(object_state[2])
```
Comment (TomAugspurger) [inline pandas/core/sparse/array.py]: Just curious: why the changes here? Did we have tests that failed without it?

File: pandas/tests/sparse/test_array.py [hunk 2]
```diff
@@ -454,6 +454,16 @@ def test_values_asarray(self):
         assert_almost_equal(self.arr.to_dense(), self.arr_data)
         assert_almost_equal(self.arr.sp_values, np.asarray(self.arr))
 
+    def test_shape(self):
+        out = SparseArray([0, 0, 0, 0, 0])
```
Comment (TomAugspurger) [inline pandas/tests/sparse/test_array.py]: Reference the github issue here as a comment.

**Review by TomAugspurger — COMMENTED** [review]
Could you also add a release note to 0.23.1

File: pandas/core/sparse/array.py (hunk 1 above)
Comment (nprad) [reply pandas/core/sparse/array.py]: Yes, `test_pickle` was failing.`shape` is used to pickle objects and the overridden `shape` does not return the correct value. Therefore, I had to set the `shape` manually.

**Comment by nprad** [comment]
Thanks for the review! Made the changes. Let me know if there is anything else that needs to be done.

File: pandas/core/sparse/array.py [hunk 3]
```diff
@@ -290,7 +290,9 @@ def __reduce__(self):
...
+        object_state[2][1] = super(SparseArray, self).shape
```
Comment (jreback) [inline pandas/core/sparse/array.py]: this is pretty obtuse, can you make this any more clear

File: pandas/tests/sparse/test_array.py [hunk 4]
```diff
@@ -454,6 +454,17 @@ def test_values_asarray(self):
         assert_almost_equal(self.arr.to_dense(), self.arr_data)
         assert_almost_equal(self.arr.sp_values, np.asarray(self.arr))
 
+    def test_shape(self):
+        # GH 21126
```
Comment (jreback) [inline pandas/tests/sparse/test_array.py]: can you add cases that have a fill_value, can you parameterize the tests. also add a test with an object array as well (e.g. steal cases from the constructor tests)

File: doc/source/whatsnew/v0.23.1.txt [hunk 5]
```diff
@@ -67,6 +67,11 @@ Categorical
 
 - Bug in :func:`pandas.util.testing.assert_index_equal` which raised ``AssertionError`` incorrectly, when comparing two :class:`CategoricalIndex` objects with param ``check_categorical=False`` (:issue:`19776`)
 
+Sparse
+^^^^^^
+
+- Bug in :attr:`SparseArray.shape` ignores ``fill_values`` (:issue:`21126`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.1.txt]: can you expand on this slightly. maybe

Bzug in SparseArray.shape which previouslvy return a shape that only included the dense values.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by nprad** [comment]
Thanks @jreback for review! Let me know if this is better

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @nprad
